pagefaultgames/pokerogue
# Added "Skip Seen Dialogues" option

## Why?
This seems to get requested a lot in the #feature-vote discord channel. It's more of a QoL for grinding players who basically click through every dialogue anyway and play at 5x speed. The dialogues can get annoying for some people especially when you F5/retry a lot at e.g. lvl195. The option only allows to skip **seen** dialogues! The dialogue key (e.g. `PGFdialogue:rival.encounter.1`) is used to mark a dialogue as seen, so the option is independent of the text of the dialogue as well as the language.

![image](https://github.com/pagefaultgames/pokerogue/assets/1838682/bb0d341e-7918-43c9-a9c8-5128dab6aabc)


## Changes
- added skip seen dialogues option
- updated ui.showDialogue to skip a dialogues if it has been seen and when the option is active.
- updated game-data.ts to save/load the key of a seen dialogue to/from local storage
- added `shouldSkipDialogue(key:string): boolean` as a helper function
- only show character sprite when **not** skipping dialogue

![image](https://github.com/pagefaultgames/pokerogue/assets/1838682/a677a840-15f8-4ddf-b7cc-08231ad81213)

## Videos & Screenshots

Here are 5 different working scenarios. Most of them showcase the first time a dialogue is seen and then the second time (this one should be skipped). The last one also shows the dialogue key working in action: changing the language doesn't effect the "has seen"-state.

**Note: This is on 5x Speed!!!**

**Scenario 1: Rival Stage 8 - Start Of Battle:**

https://github.com/pagefaultgames/pokerogue/assets/1838682/bd0ca171-d7d1-4fd8-bab1-0121fcca7824


**Scenario 2: Rival Stage 8 - End Of Battle:**

https://github.com/pagefaultgames/pokerogue/assets/1838682/8a190c11-c2b3-47d6-963d-407f9d9a5f72


**Scenario 3: Rival Stage 200 - Winning Classic Mode:**

https://github.com/pagefaultgames/pokerogue/assets/1838682/86c1ab71-942a-4e17-bf72-d63babbde4b9


**Scenario 4: Trainer Stage 5 - Start Of Battle + End Of Battle:**

https://github.com/pagefaultgames/pokerogue/assets/1838682/48aaf5b3-897a-4d68-b384-afbfa619fa66


**Scenario 5: Changing ingame language:**

https://github.com/pagefaultgames/pokerogue/assets/1838682/5fddfa37-276e-4fb6-86eb-25f5d20adef5



## How to test the changes?
Go to the game settings, find the option, activate it and go to any dialogue (you can manipulate your starting wave by changing `STARTING_WAVE_OVERRIDE` in `overrides.ts` to e.g. 5 (trainer encounter), 8 first rival encounter, 195 (the longest rival dialogue) or 201 to see the classic win dialogue). After the dialogue you can reload and check the same dialogue again. It should now get skipped. Reactivate the option and it will reappear. To check a end of battle dialogue just set `export const OPP_LEVEL_OVERRIDE: number = 1;` and `export const STARTING_LEVEL_OVERRIDE: integer = 100;` to make you OP and quickly end the fight.

## Checklist
